kin-devplatform-sdk/src/main/java/kin/devplatform/KinEcosystemInitiator.java

```diff
@@ -76,85 +80,109 @@ public void internalInit(Context context) {
-		final KinCallback<Void> loginCallback) {
+	public void externalInit(Context context, String appId, KinEnvironment environment, @NonNull SignInData signInData,
+							 final KinCallback<Void> loginCallback) {
 		if (isInitialized && isLoggedIn) {
 			fireStartCompleted(loginCallback);
 			return;
 		}
 
 		try {
-			init(context, environment);
+			init(context, appId, environment, true, signInData, loginCallback);
 		} catch (BlockchainException e) {
 			fireStartError(e, loginCallback);
-			return;
 		}
-		login(signInData, loginCallback);
 	}
 
-	private void init(Context context, KinEnvironment environment) throws BlockchainException {
+	private void init(Context context, String appId, KinEnvironment environment, boolean withLogin,
+					  SignInData signInData, KinCallback<Void> loginCallback) throws BlockchainException {
 		if (!isInitialized) {
 			ConfigurationLocal configurationLocal = ConfigurationLocal.getInstance(context);
 			if (environment != null) {
 				configurationLocal.setEnvironment(environment);
 			}
+
 			ConfigurationImpl.init(configurationLocal);
 			KinEnvironment kinEnvironment = ConfigurationImpl.getInstance().getEnvironment();
-			EventLogger eventLogger = EventLoggerImpl.getInstance();
-			final String networkUrl = kinEnvironment.getBlockchainNetworkUrl();
-			final String networkId = kinEnvironment.getBlockchainPassphrase();
+			final EventLogger eventLogger = EventLoggerImpl.getInstance();
+			final String oldNetworkUrl = kinEnvironment.getOldBlockchainNetworkUrl();
+			final String oldNetworkId = kinEnvironment.getOldBlockchainPassphrase();
+			final String newNetworkUrl = kinEnvironment.getNewBlockchainNetworkUrl();
+			final String newNetworkId = kinEnvironment.getNewBlockchainPassphrase();
 			final String issuer = kinEnvironment.getIssuer();
 
-			KinClient kinClient = new KinClient(context, new ServiceProvider(networkUrl, networkId) {
-				@Override
-				protected String getIssuerAccountId() {
-					return issuer;
-				}
-			}, KIN_ECOSYSTEM_STORE_PREFIX_KEY);
-			BlockchainSourceImpl.init(eventLogger, kinClient, BlockchainSourceLocal.getInstance(context));
+			// TODO: 01/01/2019 is it ok to use issuer like i did or like they did in the past like this:
+//			new ServiceProvider(networkUrl, networkId) {
+//				@Override
+//				protected String getIssuerAccountId() {
+//					return issuer;
+//				}
 
-			AuthRepository
-				.init(AuthLocalData.getInstance(context, executorsUtil),
-					AuthRemoteData.getInstance(executorsUtil));
+			MigrationNetworkInfo migrationNetworkInfo = new MigrationNetworkInfo(oldNetworkUrl, oldNetworkId, newNetworkUrl, newNetworkId, issuer);
+			MigrationManager migrationManager = new MigrationManager(context, appId, migrationNetworkInfo, new KinVersionProvider(), KIN_ECOSYSTEM_STORE_PREFIX_KEY);
+			try {
+				handleMigration(context, appId, eventLogger, migrationManager, withLogin, signInData, loginCallback);
```

Comment:
we have a new problem here now - `init` is a async blocking method because of migration module, the intention was to separate internal and external init by making the `init` method sync blocking method.
the potential problem is that  in case of internalInit SDK will not be ready only after migration module will do its thing, and that can lead to errors if we try to access SDK before.

Reply:
We still need to resolve it, will talk also with Segev and will see what is the best approach.